api/agent/drivers/docker/docker.go

```diff
@@ -51,17 +51,79 @@ func (r *runResult) Error() error   { return r.err }
 func (r *runResult) Status() string { return r.status }
 
 type DockerDriver struct {
-	cancel   func()
-	conf     drivers.Config
-	docker   dockerClient // retries on *docker.Client, restricts ad hoc *docker.Client usage / retries
-	hostname string
-	auths    map[string]driverAuthConfig
-	pool     DockerPool
+	cancel     func()
+	conf       drivers.Config
+	docker     dockerClient // retries on *docker.Client, restricts ad hoc *docker.Client usage / retries
+	hostname   string
+	auths      map[string]driverAuthConfig
+	pool       DockerPool
+	imageCache *Cache
 	// protects networks map
 	networksLock sync.Mutex
 	networks     map[string]uint64
 }
 
+// NewImageCleaner builds an evicter that loops every ImageCacheCleanInterval, checkes to see
+// if there is more space consumed then the maxSize allowed by configuration. If there is
+// NewImageCleaner checks the image cache for the list of evicitable images and then tries
+// in order of most evicitable to remove the image. If the disk consumption constraint is
+// satisfied the for loop breaks.
+func NewImageCleaner(context context.Context, dockerDriver *DockerDriver, maxSize int64) error {
+	opts := docker.RemoveImageOptions{}
+	duopts := docker.DiskUsageOptions{}
+	opts.Force = true
+	opts.NoPrune = false
+	opts.Context = context
+	duopts.Context = context
+	logrus.Info("Starting image cleaner")
+	ticker := time.NewTicker(dockerDriver.conf.ImageCacheCleanInterval)
+	for {
+		select {
+		case <-context.Done():
+			return nil
+		case <-ticker.C:
+			du, err := dockerDriver.docker.DiskUsage(duopts)
+			if err != nil {
+				logrus.WithError(err).Error("attempting to check disk usage")
+			}
+			if du.LayersSize > maxSize {
+				toEvict := dockerDriver.imageCache.Evictable()
+				for _, i := range toEvict {
+					err := dockerDriver.docker.RemoveImage(i.image.ID, opts)
+					if err != nil {
+						logrus.WithError(err).Errorf("Could not remove image: %v because: %v", i, err)
+					} else {
+						dockerDriver.imageCache.Remove(i.image)
```

Comment:
a bit out there, but it's possible that this ends up in inconsistencies between images that exist on disk and the image caches list of images (ie we have a fugitive image) - eg docker can succeed in removing an image but have a network error telling us about it. then we have an image in the cache that isn't there. i don't think it's the end of the world, maybe there's a pathological case in there, but it's not easy to reason about either.